File: CortexSuite/vision/benchmarks/localization/src/c/get3DGaussianProb.c
/********************************
Author: Sravanthi Kota Venkata
********************************/

#include <stdio.h>
#include <stdlib.h>
#include "localization.h"

F2D* get3DGaussianProb( F2D* data, F2D* mean, F2D* A)
{
    F2D *p, *diff, *temp1, *temp2, *mt;
    float temp;
    int n_data, n_channel;
    int i, j, k;
    F2D* t;
    float pi = 3.1412;

    n_data = data->height;
    n_channel = data->width;

    t = fSetArray(n_data, 1, 1);

    mt = fMtimes(t, mean); 
    diff = fMinus( data, mt);
    p = fSetArray(diff->height, 1, 0);

    temp = sqrt(1.0/(pow(2*pi, n_channel)));
    temp2 = randWrapper(diff->height,1);

    j = (temp2->height*temp2->width);
    for(i=0; i<j; i++)
    {
        float temp2i;
        temp2i = asubsref(temp2,i);

        temp2i = exp(-0.5*temp2i);
        asubsref(p,i) = temp2i*temp;
    }

    fFreeHandle(t);
    fFreeHandle(temp2);
    fFreeHandle(mt);
    fFreeHandle(diff);
    
    return p;
}

File: CortexSuite/vision/benchmarks/localization/src/matlab/get3DGaussianProb.m
function p=get3DGaussianProb(data, mean, A)
n_data=size(data,1);
n_channel=size(data,2);

p=zeros(n_data,1);
diff=(data)-ones(n_data,1)*mean;
detA = 1; %detA = det(A)
dotA = randWrapper(size(diff,1),1); %dotA = dot(diff*A, diff, 2)
p=sqrt(detA/((2*pi)^n_channel))*exp(-0.5*dotA);

%% KVS If the above doesnt work, try uncommenting these lines below

%%temp = (det(A)/((2*pi)^n_channel));
%temp = (1.0/((2*pi)^n_channel));
%temp1 = dot(diff*A,diff,2);
%%temp1 = rand(1000,1); 
%temp2 = exp(-0.5*temp1);
%p = sqrt(temp) * exp(temp2);
%



